wordpress-mobile/WordPress-iOS
# Likes List: filter out users already fetched

WPKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/402

 There is a scenario where multiple users might like a post/comment at the same time, and then end up split between pages of results. So we'll track which users we've already fetched for the `lastFetchedDate`, and send those to the Post and Comment services to filter out of the response so we don't get duplicates.

Also, since this is the last part of Like Notifications, the feature flag is enabled for all.

To test:
- Go to Notifications > Likes.
- Select a notification with:
  - A lot of Likes (See #16547 for a hack tip if needed.)
    - If you manually set the IDs, you'll need to do the same [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/004747e5ccdd5482a7c2e9717b892854382f19e8/WordPress/Classes/ViewRelated/Likes/LikesListController.swift#L184).
  - Likes with the same date/time. See hack tip below to muck with the dates if needed.
- Scroll to the bottom of the list so the next page loads.
- Verify there are no duplicates at the end of one page and the beginning of the next, and no users are missing.

Hack tip for dates:
In order to have multiple Likes with the same date/time, this should work. 
- In the `createNewUsers` methods in [`PostService+Likes`](https://github.com/wordpress-mobile/WordPress-iOS/blob/d943090f9236b9405ca8c6116a05c5f2892bd5e5/WordPress/Classes/Services/PostService%2BLikes.swift#L103) / [`CommentService+Likes`](https://github.com/wordpress-mobile/WordPress-iOS/blob/d943090f9236b9405ca8c6116a05c5f2892bd5e5/WordPress/Classes/Services/CommentService%2BLikes.swift#L103) when they call `createUserFrom`, manually change the `dateLiked` to the same date. For example:

```
//            remoteLikeUsers.forEach {
//                LikeUserHelper.createUserFrom(remoteUser: $0, context: derivedContext)
//            }
            
            for (index, user) in remoteLikeUsers.enumerated() {
                let endDate = remoteLikeUsers.last?.dateLiked
                
                if index == remoteLikeUsers.endIndex - 3 {
                    user.dateLiked = endDate
                }
                
                LikeUserHelper.createUserFrom(remoteUser: user, context: derivedContext)
            }
``` 

## Regression Notes
1. Potential unintended areas of impact
N/A. Feature incomplete.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
N/A. Feature incomplete.

3. What automated tests I added (or what prevented me from doing so)
N/A. Feature incomplete.

PR submission checklist:
